File: 15-3sum/3sum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
    nums.sort((a, b) => a - b);
    let ans = [];
    for(let i = 0; i < nums.length -2; i++){
        if (i > 0 && nums[i] === nums[i - 1]) continue;
            let left = i + 1, right = nums.length - 1;
            while (left < right) {
                const sum = nums[i] + nums[left] + nums[right];
                if (sum === 0) {
                    ans.push([nums[i], nums[left], nums[right]]);
                    while (nums[left] === nums[left + 1]) left++;
                    while (nums[right] === nums[right - 1]) right--;
                    left++;
                    right--;
                }
                else if (sum > 0) right--;
                else left++;
            }    
    }
    return ans;
};

File: 15-3sum/3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] arr) {
         List<List<Integer>> set = new ArrayList();
        Arrays.sort(arr);
    
        for(int i=0; i<arr.length; i++){
            if(i > 0 && arr[i] == arr[i-1]) continue;
            int left = i+1, right = arr.length-1;
            
            while(left < right){
                int sum = arr[i] + arr[left] + arr[right];
                
                if(sum == 0){
                    set.add(Arrays.asList(arr[i], arr[left++], arr[right--]));
                    
                    while(left < right && arr[left] == arr[left-1]) left++;
                    while(left < right && arr[right] == arr[right+1]) right--;
                } else if(sum < 0){
                    left++;
                } else{
                    right--;
                }
            }
        }
        
        return set;
    }
}